### PR TITLE
Add support for installing the Steam Streaming Speakers driver

### PR DESCRIPTION
**Pull request by cgutman**
## Description
This PR adds a feature to install the Steam Streaming Speakers driver if it's present on disk (Steam installed) but not already installed (if the user hasn't used Steam Remote Play). Since some users may have their own audio configuration that they don't want us messing with, I added a UI option to opt-out of this behavior.

The Steam Streaming Speakers driver allows us to support 5.1 and 7.1 surround sound on systems without a physical surround sound output. It also allows the "Play audio on host PC" option in Moonlight to correctly mute host audio.

### Screenshot
image

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -382,7 +382,11 @@ namespace config {
     true  // dwmflush
   };
 
-  audio_t audio {};
+  audio_t audio {
+    {},  // audio_sink
+    {},  // virtual_sink
+    true,  // install_steam_drivers
```
Comment (ReenigneArcher) [inline src/config.cpp:388]:
```suggestion
    false,  // install_steam_drivers
```

Default should probably be false, so we don't mess with anything until they opt in?

File: src/platform/windows/audio.cpp [hunk 2]
```diff
@@ -878,6 +862,78 @@ namespace platf::audio {
       return std::nullopt;
     }
 
+    /**
+     * @brief Installs the Steam Streaming Speakers driver, if present.
+     * @return true if installation was successful
```
Comment (ReenigneArcher) [inline src/platform/windows/audio.cpp:867]:
```suggestion
     * @return `true` if installation was successful.
```

File: src_assets/common/assets/web/config.html [hunk 3]
```diff
@@ -992,6 +1004,7 @@ <h1 class="my-4">Configuration</h1>
     "amd_vbaq": "enabled",
     "capture": "",
     "controller": "enabled",
+    "install_steam_audio_drivers": "enabled",
```
Comment (ReenigneArcher) [inline src_assets/common/assets/web/config.html:1007]:
```suggestion
    "install_steam_audio_drivers": "disabled",
```

**Review by ReenigneArcher — COMMENTED** [review]
I'll trust your c++... my biggest observation is the default value. I'm thinking this should be disabled by default to avoid installing extra drivers in cases where a user might not want it.

**Comment by cgutman** [comment]
> I'll trust your c++... my biggest observation is the default value. I'm thinking this should be disabled by default to avoid installing extra drivers in cases where a user might not want it.

I think it should be enabled by default for a few reasons:
- Steam installs it when you stream with Steam Remote Play, so there's existing precedent for installing audio drivers by default as part of game streaming.
- Host audio mute is a default enabled behavior in Moonlight, so it's likely to be used in the vast majority of streaming scenarios. Sunshine shouldn't require special configuration to work as expected (to the extent we can, of course).
- We take care to set their original audio device back to default after installing Steam Streaming Speakers to prevent changes to their default audio output as a result of the installation (Steam does this too)

For the scenario where the user doesn't want our default, I think it's much more likely that a user will recognize what's happening if they see Sunshine installing that driver and want to stop it vs. a user wondering why they pick 5.1 surround sound in Moonlight and get upmixed stereo or why Moonlight's host audio mute seems to be broken. I don't think users in the latter case would necessarily understand what's going on to try to fix it.

There seem to me to be far more users where this behavior will help than where it will hurt. If a user doesn't want it, they can simply remove the device and disable the option in Sunshine. We can also document it in the "Breaking Changes" section to ensure upgrading users adjust the option to their preference before streaming.

**Comment by ReenigneArcher** [comment]
> adjust the option to their preference before streaming

So it doesn't install the driver until a stream starts? If that's the case, that will be better than installing it once sunshine starts up.

I trust your judgement here. If it becomes an issue I can add some instructions for manually removing the driver.

**Comment by cgutman** [comment]
> So it doesn't install the driver until a stream starts? If that's the case, that will be better than installing it once sunshine starts up.

Yep, it only installs when a stream actually starts.

**Review by ReenigneArcher — DISMISSED** [review]

**Comment by cgutman** [comment]
Pushed a small fix to let us successfully install the driver in the case where there are no audio devices on the system (therefore no existing default device to restore).

**Review by ReenigneArcher — APPROVED** [review]

**Comment by FallingSnow** [comment]
When using the steam audio driver and dragging the audio slider to 0, the audio gets progressively quieter. However once I hit 0 (and the windows icon changes to muted) the output volume is maxed again. So while windows says it is muted, the client plays volume at max.

Not sure if this is a stream driver issue.

**Comment by Roshan-Localhost** [comment]
Hi @cgutman @ReenigneArcher 
For Some Reason for me the steam streaming drivers are not getting installed.
As per logs it does says that steam streaming drivers has been installed, but on device manager it doesn't show up anything. 
And the Audio is playing on both Moonlight side pc and on PC where sunshine installed.

` Info: Successfully installed Steam Streaming Speakers`

But, It does works if I Manually install the steam drivers, but it was a pain to install the steam drivers and there are now multiple copies of steam streaming speakers on my pc. 

btw I'm on windows 11.

Thanks

**Comment by Volumie** [comment]
Hi @cgutman 

For me Steaming streaming speaker is not getting installed.
But the logs does say:
`Info: Successfully installed Steam Streaming Speakers`

So Sound is getting played on both the Moonlight PC and Sunshine PC.

Thanks

**Pull request merged**